src/main/kotlin/com/manimdsl/frontend/ManimParserVisitor.kt

```diff
@@ -275,6 +275,88 @@ class ManimParserVisitor : ManimParserBaseVisitor<ASTNode>() {
         return whileStatementNode
     }
 
+    private fun visitForHeader(ctx: ForHeaderContext): Triple<DeclarationNode, ExpressionNode, AssignmentNode> {
+        return visitRange(ctx as RangeHeaderContext)
+    }
+
+    private fun visitRange(ctx: RangeHeaderContext): Triple<DeclarationNode, ExpressionNode, AssignmentNode> {
+        val lineNumber = ctx.start.line
+        val identifier = ctx.IDENT().symbol.text
+        semanticAnalyser.redeclaredVariableCheck(symbolTable, identifier, ctx)
+        val startExpr = if (ctx.begin != null) {
+            visit(ctx.begin) as ExpressionNode
+        } else {
+            NumberNode(lineNumber, 0.0)
+        }
+        val end = visit(ctx.end) as ExpressionNode
+        val change = if (ctx.delta != null) {
+            visit(ctx.delta) as ExpressionNode
+        } else {
+            NumberNode(lineNumber, 1.0)
+        }
+        semanticAnalyser.forLoopRangeUpdateNumberTypeCheck(symbolTable, change, ctx)
+        semanticAnalyser.forLoopRangeTypeCheck(symbolTable, startExpr, end, ctx)
+
+        val start = DeclarationNode(lineNumber, IdentifierNode(lineNumber, identifier), startExpr)
+        val counterType = semanticAnalyser.inferType(symbolTable, startExpr)
+        val update = if (counterType is CharType) {
+            AssignmentNode(
+                lineNumber,
+                IdentifierNode(lineNumber, identifier),
+                CastExpressionNode(
+                    lineNumber,
+                    CharType,
+                    AddExpression(
+                        lineNumber,
+                        IdentifierNode(lineNumber, identifier),
+                        change
+                    )
+                )
+            )
+        } else{
+            AssignmentNode(
+                lineNumber,
+                IdentifierNode(lineNumber, identifier),
+                AddExpression(
+                    lineNumber,
+                    IdentifierNode(lineNumber, identifier),
+                    change
+                )
+            )
+        }
+        symbolTable.addVariable(identifier, IdentifierData(counterType))
+        return Triple(start, end, update)
+    }
+
+    override fun visitForStatement(ctx: ForStatementContext): ASTNode {
+        val prevInLoop = inLoop
+        val forScope = if (!prevInLoop) {
```

Comment:
Would you not want to enter a new scope regardless of whether you were already in a loop or not?

Reply:
Yeah I did that initially but then I thought we don't want the user to be able to redefine either the counter `i` or any variable within nested loops, so I thought it would be better if I keep them all in the same scope?

Reply:
Okay, I understand - what about for a program like this?

```
let x = 0;
let y = 0;
for (x in range(10)) {
    for (y in range (5)) {
        y = y + 1;
    }
    x = x + 1;
}
```

Would `y` be accessible in the outer for loop since they're treated as being in the same scope?

Reply:
The first problem is now I don't think I allow y to be declared as the identifier for the for loop counter if y is variable declared outside the loop, but if we want to be able to do that I can look into it.

and regarding `z`, do we want to be able to access it in the outer loop or not? I can change things and test things depending on what we want

Reply:
> The first problem is now I don't think I allow y to be declared as the identifier for the for loop counter if y is variable declared outside the loop, but if we want to be able to do that I can look into it.

I just checked what Python does - if `y` is declared before the for header, it allows it semantically but ignores the previous declaration when going into the loop. We can leave it for now and discuss later with the rest of the group.

Reply:
yeah and I think when I tried nested for loops in python, it also allows reusing the `i` in nested for loop and redeclaring variables but I don't think we wanted that? So yeah, maybe can discuss with the rest of the group

Reply:
> and regarding z, do we want to be able to access it in the outer loop or not? I can change things and test things depending on what we want

I think we wouldn't want to access `z` in the outer loop. Perhaps instead of treating the scopes as the same, we can enter a new scope each time we enter a new for loop and keep track of the current identifiers being used to iterate over.

So if we had something of the form:

```
for i in range(5) {
  for j in range(3) {
    # ...
  }
}
```

we can treat `i` and `j` differently. So if you're within a for loop, and the user tries to reassign `i` or `j` to something, we throw a semantic error.